A40-csds-support.md

```diff
@@ -0,0 +1,328 @@
+this information is static and may require extra logic to integrate into the
+CSDS service.
+
+For Java and Go, it’s straightforward that this feature can be implemented
+bottom-up in the same language, and it’s possible to avoid an additional copy of
+the message itself. But Core needs to transport the collected xDS configs across
+the language boundary. The API needs to convert proto messages into C-compatible
+types (e.g. `char *`, either bytes or JSON).
+
+
+#### Detail: No xDS v2 Support
+
+CSDS was meant to be a RPC service made for xDS management servers before this
+proposal. But the service itself has the potential to serve client status on xDS
+clients. However, during development, we found several constraints about the
+existing service protocol, hence we merged several updates (see above) to
+improve the CSDS service to meet our standard. The updates are made to xDS v3
+and xDS v4 (alpha) only, since xDS v2 is in deprecated state. This doc proposes
+to not support xDS v2 for CSDS.
+
+#### Detail: Expose xDS Config As Is
+
+To date, gRPC supports the majority of xDS config but not all. Alternatively, we
+could only dump the config that gRPC understands. However, doing so will create
+a behavior difference between the gRPC and the control plane. It might be
+confusing to users that some fields are missing. Also, during the configuration
+dump, Envoy also dumps its entire configuration even if it doesn’t understand
+some of the configuration fields (think of fields for Envoy extensions). So, for
+correctness and simplicity, we should **cache and expose the xDS config the way
+gRPC receives them**.
+
+#### Detail: Cache Lifecycle
+
+gRPC doesn't use xDS messages directly, but interprets them into language-native
+class/struct. The lifecycle of the cached xDS messages should be identical to
+the interpreted structure in each stack, so there is no memory management logic
+change needed.
+
+#### Detail: Generated Node Information
+
+gRPC loads Node information from a [bootstrap
+file](https://github.com/grpc/proposal/blob/master/A27-xds-global-load-balancing.md#xdsclient-and-bootstrap-file),
+which includes the identification that will be used in the control plane. gRPC's
+xDS client also generates `user_agent_name`, `user_agent_version`, and
+`client_features` before sending the `Node` information to the control plane.
+These information can be very helpful to debugging. This doc recommends the
+implementation to provide the `ClientStatusResponse.ClientConfig.node` field
+with the `Node` information that the control plane will receive. Here is an
+example of filled `Node`:
+
+```json
+{
+  "node": {
+    "id": "c591240a-b3b6-4761-aada-3fa43ec6a852",
+    "cluster": "cluster",
+    "metadata": {
+      "...": "..."
+    },
+    "locality": {
+      "zone": "zone"
+    },
+    "userAgentName": "gRPC Java",
+    "userAgentVersion": "1.35.1-SNAPSHOT",
+    "clientFeatures": ["envoy.lb.does_not_support_overprovisioning"]
+  }
+}
+```
```

Comment:
The requirement of Node information is updated.
CC @sergiitk @menghanl

Reply:
Thanks for capturing this! Glad we've figured it out.

Reply:
As I mentioned above, I don't see any compelling reason to bother with populating this field.

Reply:
@markdroth @lidizheng @menghanl and I discussed that in a chat. To summarize:
1) Parity with [`istioctl proxy-config bootstrap `](https://istio.io/latest/docs/ops/diagnostic-tools/proxy-cmd/#inspecting-bootstrap-configuration)
2) Some fields are generated after reading the bootstrap and sent to the control plane in requests: `user_agent_name`, `user_agent_version`, and, more importantly `clientFeatures`. Exposing them can be helpful for debugging. Note this is different from regular CSDS behavior to dump configs as is. The argument for treating the node info differently is that it's loaded from bootstrap and _sent_ to the control plane, as opposed to it's _received_ from the control plane

Reply:
If there's desire to expose the client's node information for debugging purposes, then that's a perfectly reasonable use-case, and we should design something to address it.  However, CSDS was not designed to solve that problem, and IMHO it would be a very clunky way to solve that problem.  If the mechanism we provide for accessing the bootstrap info is for a client to query CSDS, that means that a client that wants only the bootstrap info has no way to get it that does not also send the entire set of known xDS resources.  That's sending a potentially very large amount of completely unnecessary data.

If we want a way to expose the client's bootstrap info for debugging purposes, let's add a separate RPC service for that.  I don't think it should be part of this design.

Reply:
I think I might be missing your point. Could you please clarify what data (if any) should be in [`ClientStatusResponse. ClientConfig.node`](https://github.com/envoyproxy/envoy/blob/98ac59243cf0c2960f267664da88c08f8b9641d1/api/envoy/service/status/v3/csds.proto#L126-L127)?

Reply:
> it would be a very clunky way to solve that problem

I agree the Node information will be sent along other xDS config dumps. And users might not need them all the time. But the same argument could apply to all xDS configs, like users just updated listeners and would like to see the listeners, and we are sending routes, clusters, endpoints in the CSDS response.

When the xDS implements get popularity soon, we might have users posting issues, and with the node information we could tell which gRPC language/release it is using, and (possibly) which control plane it is trying to talk to.

Reply:
@sergiitk 
> Could you please clarify what data (if any) should be in [`ClientStatusResponse. ClientConfig.node`](https://github.com/envoyproxy/envoy/blob/98ac59243cf0c2960f267664da88c08f8b9641d1/api/envoy/service/status/v3/csds.proto#L126-L127)?

What I am proposing is that we do not populate that field at all.

@lidizheng 
> But the same argument could apply to all xDS configs, like users just updated listeners and would like to see the listeners, and we are sending routes, clusters, endpoints in the CSDS response.

I don't see what that statement has to do with the argument I'm making here.  It's true that CSDS does not provide the ability to filter which set of configs are returned, but that's equally true whether CSDS is being served by the control plane or the client.  (And this would be trivial to add if someone needed it just by adding a resource type filter to the CSDS request proto.)

The argument I am making here is that the reason that the node field exists in the CSDS response is that CSDS was designed to be served by control planes, and the control plane will be returning data for multiple clients and needs a way to indicate which client each set of data is relevant to.  But when CSDS is sent on a client, all of the data is by definition for the same client, so there's no reason to populate this field.  In other words, the node field needs to be populated only when the CSDS server is a control plane; it is simply not relevant when the CSDS server is a client.

> When the xDS implements get popularity soon, we might have users posting issues, and with the node information we could tell which gRPC language/release it is using, and (possibly) which control plane it is trying to talk to.

IMHO, the fact that this could conceivably be useful to someone someday is not enough justification to add it.  If that was a good enough justification, why isn't it standard practice for every API to include the gRPC language and release?  Why doesn't channelz include this info?  Why doesn't every single RPC service designed by applications provide this info?

In case it's not clear, the answer is that if it's not what a given API is designed to do, it shouldn't be part of that API.  The microservices approach means that each API should provide only the data that is necessary for the problem that service is designed to solve.  If that wasn't the case, then all APIs would be huge, because everyone would throw everything they could think of into every one of them.

CSDS is designed to show the set of resources seen by the client, not the client's node information.  If we want something to expose the client's node information, that's IMHO a separate use-case, and we should design a separate service to solve that problem.

In any case, I don't think this is important enough to justify spending more time debating it, so I'll leave it up to you.